cpp/src/arrow/chunk_resolver.cc

```diff
@@ -0,0 +1,84 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/chunk_resolver.h"
+
+#include <algorithm>
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "arrow/array.h"
+#include "arrow/record_batch.h"
+
+namespace arrow {
+namespace internal {
+
+namespace {
+static inline std::vector<int64_t> MakeArraysOffsets(const ArrayVector& chunks) {
+  std::vector<int64_t> offsets(chunks.size());
```

Comment:
`chunks.size() + 1` would be better, otherwise the ending `push_back` may trigger a realloc.

Reply:
Yep, allocating in constructor and using index for last element seems to run faster ([quick-bench](https://quick-bench.com/q/5_7pn-NOvLsI0YXLrFjff7DO1lk)).

Reply:
[It seems using `offsets.reserve(chunks.size() + 1)` performs better.](https://quick-bench.com/q/3q73l1FwW_HJsCaID8GCeMV42Ck)

Reply:
Your `reserve_push_back` doesn't look right, it constructs a 1-element vector.

Reply:
Oops! side-effect of using translating for-loop to std::transform.